firedrake/dmplex.pyx

```diff
@@ -2103,3 +2103,36 @@ def halo_end(PETSc.SF sf, dat, MPI.Datatype dtype, reverse):
         CHKERR(PetscSFBcastEnd(sf.sf, dtype.ob_mpi,
                                <const void *>buf.data,
                                <void *>buf.data))
+
+
+def to_petsc_numbering(PETSc.Vec vec, V):
+    """Reorder a PETSc Vec corresponding to a Firedrake Function 
+       w.r.t. to initial PETSc numbering
+
+    :arg vec: the PETSc Vec to reorder
+    :arg V: the FunctionSpace of the function from which vec comes from
+
+    :ret out: a copy of vec ordered with PETSc numbering
+    """
+    cdef int           dim, idx, start, end, p, dof, off, d, k
+    cdef PETSc.Vec     out
+    cdef PETSc.Section section
+    cdef np.ndarray    varray, oarray
+
+    section = V.dm.getDefaultGlobalSection()
+    out = vec.duplicate()
+    varray = vec.array_r
+    oarray = out.array
+    dim = V.dim
+    idx = 0
+    start, end = vec.getOwnershipRange()
+    for p in range(*section.getChart()):
+        dof = section.getDof(p)
+        if dof > 0:
+            off = section.getOffset(p)
+            if off >= 0:
+                for d in range(dof):
+                    for k in range(dim):
+                        oarray[idx] = varray[dim*off + d*off + k - start]
+                        idx += 1
```

Comment:
If you want this to compile down to C code, you need to do:

```python
CHKERR(PetscSectionGetDof(section.sec, p, &dof))
CHKERR(PetscSectionGetOffset(section.sec, p, &off))
```

In the appropriate places.

Also, you should type-annotate the numpy arrays a `np.ndarray[PetscReal, mode="c", ndim=1]`

You can run `cython -a dmplex.pyx` and open the resulting `dmplex.html`.  You want the inner loop of your function to not have any yellow background.  Look at how the other functions in this file are annotated.